File: 2019/python/computer.py

from collections import defaultdict


MODE_POSITION = 0
MODE_IMMEDIATE = 1
MODE_RELATIVE = 2


def read_program(file_path):
    return list(map(int, open(file_path, 'r').read().strip().split(',')))


class IntcodeError(Exception):
    pass


class IntcodeComputer(object):
    def __init__(self, program, console):
        memory = defaultdict(int)
        for i, n in enumerate(program):
            memory[i] = n
        self.memory = memory
        self.console = console
        self.pointer = 0
        self.relative_base = 0


    def run(self):
        while True:
            opcode = self.memory[self.pointer]
            # print('opcode', opcode)
            opcode_n = opcode % 100
            mode1 = (opcode // 100) % 10
            mode2 = (opcode // 1000) % 10
            mode3 = (opcode // 10000) % 10
            if opcode_n == 1:
                self._add(mode1, mode2, mode3)
            elif opcode_n == 2:
                self._multiply(mode1, mode2, mode3)
            elif opcode_n == 3:
                self._input(mode1, mode2, mode3)
            elif opcode_n == 4:
                self._output(mode1, mode2, mode3)
            elif opcode_n == 5:
                self._jump_if_true(mode1, mode2, mode3)
            elif opcode_n == 6:
                self._jump_if_false(mode1, mode2, mode3)
            elif opcode_n == 7:
                self._less_than(mode1, mode2, mode3)
            elif opcode_n == 8:
                self._equals(mode1, mode2, mode3)
            elif opcode_n == 9:
                self._adjust_relative_base(mode1, mode2, mode3)
            elif opcode_n == 99:
                break
            else:
                raise IntcodeError


    def _add(self, mode1, mode2, mode3):
        '''Opcode 1'''
        p1 = self.memory[self.pointer + 1]
        p2 = self.memory[self.pointer + 2]
        p3 = self.memory[self.pointer + 3]
        n1 = self._get(p1, mode1)
        n2 = self._get(p2, mode2)
        n = n1 + n2
        if mode3 == MODE_RELATIVE:
            p3 += self.relative_base
        self.memory[p3] = n
        self.pointer += 4


    def _multiply(self, mode1, mode2, mode3):
        '''Opcode 2'''
        p1 = self.memory[self.pointer + 1]
        p2 = self.memory[self.pointer + 2]
        p3 = self.memory[self.pointer + 3]
        n1 = self._get(p1, mode1)
        n2 = self._get(p2, mode2)
        n = n1 * n2
        if mode3 == MODE_RELATIVE:
            p3 += self.relative_base
        self.memory[p3] = n
        self.pointer += 4


    def _input(self, mode1, mode2, mode3):
        '''Opcode 3'''
        assert mode2 == 0 and mode3 == 0
        # n = int(sys.stdin.readline().strip())
        n = int(self.console.readline().strip())
        p1 = self.memory[self.pointer + 1]
        if mode1 == MODE_RELATIVE:
            p1 += self.relative_base
        self.memory[p1] = n
        self.pointer += 2


    def _output(self, mode1, mode2, mode3):
        '''Opcode 4'''
        assert mode2 == 0 and mode3 == 0
        p1 = self.memory[self.pointer + 1]
        n = self._get(p1, mode1)
        # sys.stdout.write(f'{n}\n')
        self.console.write(f'{n}')
        self.pointer += 2


    def _jump_if_true(self, mode1, mode2, mode3):
        '''Opcode 5'''
        assert mode3 == 0
        p1 = self.memory[self.pointer + 1]
        n = self._get(p1, mode1)
        if n != 0:
            p2 = self.memory[self.pointer + 2]
            new_p = self._get(p2, mode2)
            self.pointer = new_p
        else:
            self.pointer += 3


    def _jump_if_false(self, mode1, mode2, mode3):
        '''Opcode 6'''
        assert mode3 == 0
        p1 = self.memory[self.pointer + 1]
        n = self._get(p1, mode1)
        if n == 0:
            p2 = self.memory[self.pointer + 2]
            new_p = self._get(p2, mode2)
            self.pointer = new_p
        else:
            self.pointer += 3


    def _less_than(self, mode1, mode2, mode3):
        '''Opcode 7'''
        p1 = self.memory[self.pointer + 1]
        p2 = self.memory[self.pointer + 2]
        p3 = self.memory[self.pointer + 3]
        n1 = self._get(p1, mode1)
        n2 = self._get(p2, mode2)
        if mode3 == MODE_RELATIVE:
            p3 += self.relative_base
        self.memory[p3] = 1 if n1 < n2 else 0
        self.pointer += 4


    def _equals(self, mode1, mode2, mode3):
        '''Opcode 8'''
        p1 = self.memory[self.pointer + 1]
        p2 = self.memory[self.pointer + 2]
        p3 = self.memory[self.pointer + 3]
        n1 = self._get(p1, mode1)
        n2 = self._get(p2, mode2)
        if mode3 == MODE_RELATIVE:
            p3 += self.relative_base
        self.memory[p3] = 1 if n1 == n2 else 0
        self.pointer += 4


    def _adjust_relative_base(self, mode1, mode2, mode3):
        '''Opcode 9'''
        assert mode2 == 0 and mode3 == 0
        p1 = self.memory[self.pointer + 1]
        n1 = self._get(p1, mode1)
        self.relative_base += n1
        self.pointer += 2


    def _get(self, p, mode):
        if mode == MODE_POSITION:
            return self.memory[p]
        if mode == MODE_IMMEDIATE:
            return p
        if mode == MODE_RELATIVE:
            return self.memory[p + self.relative_base]


class Console(object):
    def __init__(self, buffer):
        self.pointer = 0
        self.buffer = [str(b) for b in buffer]

    def readline(self):
        s = self.buffer[self.pointer]
        self.pointer += 1
        return s

    def write(self, s):
        self.buffer.append(str(s))


def test1():
    '''Day 9 test'''
    program = [109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99]
    c = Console([])
    ic = IntcodeComputer(program, c)
    ic.run()
    program1 = [int(b) for b in c.buffer]
    assert program1 == program


def test2():
    '''Day 9 test'''
    program = [1102,34915192,34915192,7,4,7,99,0]
    c = Console([])
    ic = IntcodeComputer(program, c)
    ic.run()
    x = c.readline()
    assert len(x) == 16


def test3():
    '''Day 9 test'''
    program = [104,1125899906842624,99]
    c = Console([])
    ic = IntcodeComputer(program, c)
    ic.run()
    x = c.readline()
    assert x == '1125899906842624'


def main():
    test1()
    test2()
    test3()


if __name__ == '__main__':
    main()

File: 2019/swift/AdventOfCode2019a/AdventOfCode2019a/Computer.swift
//
//  Computer.swift
//  AdventOfCode2019a
//
//  Created by gary on 14/01/2023.
//

import Foundation

typealias IntCode = Array<Int>

class Computer {
    enum Instruction {
        enum Mode: Int {
            case position = 0
            case immediate = 1
        }

        case adds(Mode, Mode, Mode)
        case multiplies(Mode, Mode, Mode)
        case input(Mode)
        case output(Mode)
        case jumpIfTrue(Mode, Mode)
        case jumpIfFalse(Mode, Mode)
        case lessThan(Mode, Mode, Mode)
        case equals(Mode, Mode, Mode)
        case halt
    }


    private(set) var memory: IntCode = []
    private var pointer = 0

    var input: [Int] = []
    var output: [Int] = []


    func load(program: IntCode) {
        self.memory = program
        self.pointer = 0
        self.input = []
        self.output = []
    }


    func run() {
        while true {
            switch Instruction(n: memory[pointer]) {
            case .adds(let m1, let m2, let m3):
                adds(mode1: m1, mode2: m2, mode3: m3)
            case .multiplies(let m1, let m2, let m3):
                multiplies(mode1: m1, mode2: m2, mode3: m3)
            case .input(let m1):
                let i = input.popLast()!
                input_(mode1: m1, input: i)
            case .output(let m1):
                let o = output_(mode1: m1)
                output.append(o)
            case .jumpIfTrue(let m1, let m2):
                jumpIfTrue(mode1: m1, mode2: m2)
            case .jumpIfFalse(let m1, let m2):
                jumpIfFalse(mode1: m1, mode2: m2)
            case .lessThan(let m1, let m2, let m3):
                lessThan(mode1: m1, mode2: m2, mode3: m3)
            case .equals(let m1, let m2, let m3):
                equals(mode1: m1, mode2: m2, mode3: m3)
            case .halt:
                return
            }
        }
    }
}


// MARK: - Private

private extension Computer {
    func adds(mode1: Instruction.Mode, mode2: Instruction.Mode, mode3: Instruction.Mode) {
        assert(mode3 == .position)
        let (p1, p2, p3) = get3()
        let (n1, n2) = get(p1: p1, m1: mode1, p2: p2, m2: mode2)

        let n = n1 + n2
        memory[p3] = n
        pointer += 4
    }


    func multiplies(mode1: Instruction.Mode, mode2: Instruction.Mode, mode3: Instruction.Mode) {
        assert(mode3 == .position)
        let (p1, p2, p3) = get3()
        let (n1, n2) = get(p1: p1, m1: mode1, p2: p2, m2: mode2)

        let n = n1 * n2
        memory[p3] = n
        pointer += 4
    }


    func input_(mode1: Instruction.Mode, input: Int) {
        assert(mode1 == .position)
        let p1 = memory[pointer + 1]
        memory[p1] = input
        pointer += 2
    }


    func output_(mode1: Instruction.Mode) -> Int {
        let p1 = memory[pointer + 1]
        let n1 = mode1 == .position ? memory[p1] : p1
        pointer += 2
        return n1
    }


    func jumpIfTrue(mode1: Instruction.Mode, mode2: Instruction.Mode) {
        let (p1, p2) = get2()
        let (n1, n2) = get(p1: p1, m1: mode1, p2: p2, m2: mode2)

        if n1 != 0 {
            pointer = n2
        } else {
            pointer += 3
        }
    }


    func jumpIfFalse(mode1: Instruction.Mode, mode2: Instruction.Mode) {
        let (p1, p2) = get2()
        let (n1, n2) = get(p1: p1, m1: mode1, p2: p2, m2: mode2)

        if n1 == 0 {
            pointer = n2
        } else {
            pointer += 3
        }
    }


    func lessThan(mode1: Instruction.Mode, mode2: Instruction.Mode, mode3: Instruction.Mode) {
        assert(mode3 == .position)
        let (p1, p2, p3) = get3()
        let (n1, n2) = get(p1: p1, m1: mode1, p2: p2, m2: mode2)

        if n1 < n2 {
            memory[p3] = 1
        } else {
            memory[p3] = 0
        }
        pointer += 4
    }


    func equals(mode1: Instruction.Mode, mode2: Instruction.Mode, mode3: Instruction.Mode) {
        assert(mode3 == .position)
        let (p1, p2, p3) = get3()
        let (n1, n2) = get(p1: p1, m1: mode1, p2: p2, m2: mode2)

        if n1 == n2 {
            memory[p3] = 1
        } else {
            memory[p3] = 0
        }
        pointer += 4
    }


    func get2() -> (Int, Int) {
        (memory[pointer + 1], memory[pointer + 2])
    }


    func get3() -> (Int, Int, Int) {
        (memory[pointer + 1], memory[pointer + 2], memory[pointer] + 3)
    }


    func get(p1: Int, m1: Instruction.Mode, p2: Int, m2: Instruction.Mode) -> (Int, Int) {
        let n1 = m1 == .position ? memory[p1] : p1
        let n2 = m2 == .position ? memory[p2] : p2
        return (n1, n2)
    }
}


extension Computer.Instruction {
    init(n: Int) {
        let opcode = n % 100
        switch opcode {
        case 1:
            self = .adds(mode(n, 100), mode(n, 1000), mode(n, 10000))
        case 2:
            self = .multiplies(mode(n, 100), mode(n, 1000), mode(n, 10000))
        case 3:
            self = .input(mode(n, 100))
        case 4:
            self = .output(mode(n, 100))
        case 5:
            self = .jumpIfTrue(mode(n, 100), mode(n, 1000))
        case 6:
            self = .jumpIfFalse(mode(n, 100), mode(n, 1000))
        case 7:
            self = .lessThan(mode(n, 100), mode(n, 1000), mode(n, 10000))
        case 8:
            self = .equals(mode(n, 100), mode(n, 1000), mode(n, 10000))
        case 99:
            self = .halt
        default:
            fatalError()
        }
    }
}


private func mode(_ n: Int, _ d: Int) -> Computer.Instruction.Mode {
    Computer.Instruction.Mode(rawValue: (n / d) % 10)!
}
